src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyStatus.cs

```diff
@@ -0,0 +1,299 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    /// <summary>
+    /// Represents the concurrency status for a function.
+    /// </summary>
+    public class ConcurrencyStatus
+    {
+        internal const int NextStatusDelayDefaultSeconds = 1;
+        internal const int DefaultFailedAdjustmentQuietWindowSeconds = 30;
+        internal const int AdjustmentRunWindowSeconds = 10;
+        internal const int DefaultMinAdjustmentFrequencySeconds = 5;
+        internal const int MaxAdjustmentDelta = 5;
+
+        private readonly ConcurrencyManager _concurrencyManager;
+
+        private object _syncLock = new object();
+        private int _adjustmentRunDirection;
+        private int _adjustmentRunCount;
+
+        /// <summary>
+        /// Constructs a new instance.
+        /// </summary>
+        /// <param name="functionId">The function ID this status is for.</param>
+        /// <param name="concurrencyManager">The <see cref="ConcurrencyManager"/>.</param>
+        public ConcurrencyStatus(string functionId, ConcurrencyManager concurrencyManager)
+        {
+            if (string.IsNullOrEmpty(functionId))
+            {
+                throw new ArgumentNullException(nameof(functionId));
+            }
+            if (concurrencyManager == null)
+            {
+                throw new ArgumentNullException(nameof(concurrencyManager));
+            }
+
+            FunctionId = functionId;
+            _concurrencyManager = concurrencyManager;
+            CurrentConcurrency = 1;
+            OutstandingInvocations = 0;
+            LastConcurrencyAdjustmentTimestamp = DateTime.UtcNow;
+            MaxConcurrentExecutionsSinceLastAdjustment = 0;
+            _adjustmentRunDirection = 0;
+        }
+
+        /// <summary>
+        /// Gets the function ID this status is for.
+        /// </summary>
+        public string FunctionId { get; }
+
+        /// <summary>
+        /// Gets the recommended amount of new work this function can safely process.
+        /// When throttling is enabled, this may return 0 meaning no new work should be
+        /// started.
+        /// </summary>
+        public int FetchCount
```

Comment:
Open to naming suggestions on this.

Reply:
Hmmm, not sure about that one either :)

Reply:
The reason I have issue with `FetchCount` is that function executions don't always correspond to a fetch operation, and really what we're expressing here is how many executions are startable. What about `AvailableWorkCount` or `StartableExecutionCount`?

Reply:
How about **AvailableInvocationCount**? The "invocation" term aligns with the OutstandingInvocations member on this type. So AvailableInvocationCount is the delta between CurrentConcurrency and OutstandingInvocations.

Reply:
Renamed to AvailableInvocationCount